cogeo_mosaic/backend/base.py

```diff
@@ -0,0 +1,70 @@
+"""cogeo_mosaic.backend.base: base Backend class."""
+
+import abc
+from contextlib import AbstractContextManager
+from typing import Dict, Optional
+
+from cogeo_mosaic import version as mosaic_version
+from cogeo_mosaic.backend.utils import get_hash
+from cogeo_mosaic.utils import create_mosaic
+
+
+class BaseBackend(AbstractContextManager):
+    mosaic_def: Optional[Dict]
+
+    @abc.abstractmethod
+    def __init__(self, *args, **kwargs):
+        """Connect to backend"""
+
+    def __enter__(self):
+        """Support using with Context Managers"""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers"""
+
+    def metadata(self) -> Dict:
+        """Retrieve Mosaic metadata
+
+        Returns
+        -------
+        MosaicJSON as dict without `tiles` key.
+        """
+        if self.mosaic_def is None:
+            return {}
+
+        return {k: v for k, v in self.mosaic_def.items() if k != "tiles"}
+
+    @abc.abstractmethod
+    def tile(self, x: int, y: int, z: int):
+        """Retrieve assets for tile."""
+
+    @abc.abstractmethod
+    def point(self, lng: float, lat: float):
+        """Retrieve assets for point."""
+
+    def create(self, *args, **kwargs):
```

Comment:
> That being said, the process to create a mosaicJSON is pretty similar once you have a list of assets

That's why I have `create` as a non-`abstractmethod` and `upload` as the `abstractmethod`. `create` is the same for all backends once you have the list of URLs. Right now `create` calls `self.upload` automatically, but I could change that to `return`. In my current terminology, I think of `create` as the method to take the urls and create a MosaicJSON, and `upload` as taking a pre-formed MosaicJSON and putting it on the backend.

> I don't think that the base backend would specify how a child class creates a mosaicjson since that is specific to the particular backend.

My thought process is that no matter what backend you'll end up uploading to, an intermediate `dict` representation of a MosaicJSON is ideal. This way each backend's `upload` method can always assume it takes in a fully-formed MosaicJSON.

Reply:
> Personally I prefer the pattern of attaching helper functions which are used by children classes as methods of the base class (self._create_mosaic) but I'm not opposed to leaving it as is.

Isn't that basically what I'm doing?

Reply:
> In my current terminology, I think of create as the method to take the urls and create a MosaicJSON

But not every backend will use the logic defined in `cogeo_mosaic.utils.create_mosaic`.  For example if you have a PostGIS database of STAC Items, you don't need to read the footprint of each URL because you already have that information from querying the database.

> but I could change that to return

I like this idea, its best if each method does one thing.

Reply:
I removed this entirely, as Mosaic creation should happen entirely outside of the backends. Specifically I now require a Mosaic to exist in `__init__`. I think that if one can't be fetched or isn't provided with the right fields an error will be produced, either from failing to fetch or through pydantic checking